services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/AttachmentServiceImpl.kt

```diff
@@ -0,0 +1,110 @@
+package fi.hel.haitaton.hanke.attachment
+
+import fi.hel.haitaton.hanke.HankeNotFoundException
+import fi.hel.haitaton.hanke.HankeRepository
+import fi.hel.haitaton.hanke.currentUserId
+import java.time.OffsetDateTime
+import java.util.UUID
+import org.springframework.transaction.annotation.Transactional
+import org.springframework.web.multipart.MultipartFile
+
+private val supportedFiletypes =
+    mapOf(
+        "application/pdf" to setOf("pdf"),
+        "application/msword" to setOf("docx", "doc"),
+        "image/jpeg" to setOf("jpg", "jpeg"),
+        "image/png" to setOf("png"),
+        "image/vnd.dwg" to setOf("dwg", "dws")
+    )
+
+private const val FILESIZE_MAXIMUM_MEGABYTES = 10
+
+class AttachmentServiceImpl(
+    val hankeRepository: HankeRepository,
+    val hankeAttachmentRepository: HankeAttachmentRepository,
+) : AttachmentService {
+
+    fun getExtension(filename: String): String {
+        return filename.split(".").last().trim()
+    }
+
+    fun sanitize(filename: String): String {
+        return Regex("[^0-9a-zA-ZäåöÄÖÅ.\\s]").replace(filename.trim(), "")
```

Comment:
This will mangle many completely valid filenames.
- Castréninkatu will become Castrninkatu
- mannerheimintie-8-liikennesuunnitelma.pdf will be squashed to mannerheimintie8liikennesuunnitelma.pdf
- valtakirja_Patrik_Floréen.pdf will be valtakirjapatrikfloren.pdf

This will, however, keep all line breaks and tab symbols.

Reply:
Basically what we want is this:

https://cheatsheetseries.owasp.org/cheatsheets/File_Upload_Cheat_Sheet.html

Filename Sanitization
Filenames can endanger the system in multiple ways, either by using non acceptable characters, or by using special and restricted filenames. For Windows, refer to the following [MSDN guide](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#naming-conventions). For a wider overview on different filesystems and how they treat files, refer to [Wikipedia's Filename page](https://en.wikipedia.org/wiki/Filename).

In order to avoid the above mentioned threat, creating a random string as a file-name, such as generating a UUID/GUID, is essential. If the filename is required by the business needs, proper input validation should be done for client-side (e.g. active content that results in XSS and CSRF attacks) and back-end side (e.g. special files overwrite or creation) attack vectors. Filename length limits should be taken into consideration based on the system storing the files, as each system has its own filename length limit. If user filenames are required, consider implementing the following:

Implement a maximum length
Restrict characters to an allowed subset specifically, such as alphanumeric characters, hyphen, spaces, and periods
If this is not possible, block-list dangerous characters that could endanger the framework and system that is storing and using the files.

Reply:
Using a regex is not a good idea and the only guaranteed secure solution I can think of is that we generate the file name.

Reply:
Any opinions on this?

Reply:
If we want to verify and keep the original file name then we can do something like this: https://www.baeldung.com/java-validate-filename

Reply:
```
/**
 * Checks file name length and creates a path object from file name. Invalid name throws
 * AttachmentUploadException.
 */
fun validateFileName(fileName: String) {
    if (fileName.length > 128) {
        throw AttachmentUploadException("File name too long. Maximum is 128.")
    }
    try {
        Paths.get(fileName)
    } catch (e: InvalidPathException) {
        throw AttachmentUploadException("File name is not valid.")
    }
}
```

How about something like this?

Reply:
I decided to change approach quite a bit. At least I was not able to get any errors in the approach I suggested.

There is now a separate attachmentvalidator which takes care of the initial checks before virus scan.

Reply:
👍 for separate validators.

With the current checks, filenames shouldn't cause problems for our users.

I'm not quite sure about throwing a general exception every time there's something slightly wrong with the filename, but I guess this is something we'll get feedback on if it's troublesome.